spark/src/main/scala/geotrellis/spark/costdistance/IterativeCostDistance.scala

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright 2017 Azavea
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package geotrellis.spark.costdistance
+
+import geotrellis.proj4.LatLng
+import geotrellis.raster._
+import geotrellis.raster.costdistance.SimpleCostDistance
+import geotrellis.raster.rasterize.Rasterizer
+import geotrellis.spark._
+import geotrellis.spark.tiling._
+import geotrellis.util._
+import geotrellis.vector._
+
+import org.apache.log4j.Logger
+import org.apache.spark.rdd.RDD
+import org.apache.spark.SparkContext
+import org.apache.spark.storage.StorageLevel
+import org.apache.spark.util.AccumulatorV2
+
+import scala.collection.mutable
+
+
+/**
+  * This Spark-enabled implementation of the standard cost-distance
+  * algorithm mentioned in the "previous work" section of [1] is
+  * "heavily inspired" by the MrGeo implementation [2] but does not
+  * share any code with it.
+  *
+  * 1. Tomlin, Dana.
+  *    "Propagating radial waves of travel cost in a grid."
+  *    International Journal of Geographical Information Science 24.9 (2010): 1391-1413.
+  *
+  * 2. https://github.com/ngageoint/mrgeo/blob/0c6ed4a7e66bb0923ec5c570b102862aee9e885e/mrgeo-mapalgebra/mrgeo-mapalgebra-costdistance/src/main/scala/org/mrgeo/mapalgebra/CostDistanceMapOp.scala
+  */
+object IterativeCostDistance {
+
+  type KeyCostPair = (SpatialKey, SimpleCostDistance.Cost)
+  type Changes = mutable.ArrayBuffer[KeyCostPair]
+
+  val logger = Logger.getLogger(IterativeCostDistance.getClass)
+
+  /**
+    * An accumulator to hold lists of edge changes.
+    */
+  class ChangesAccumulator extends AccumulatorV2[KeyCostPair, Changes] {
+    private val list: Changes = mutable.ArrayBuffer.empty
+
+    def copy: ChangesAccumulator = {
+      val other = new ChangesAccumulator
+      other.merge(this)
+      other
+    }
+    def add(pair: KeyCostPair): Unit = {
+      this.synchronized { list.append(pair) }
```

Comment:
I totally forgot about it, but @echeipesh noticed these locks. @jamesmcclain, have you considered any java concurrent collection usage here?